### PR TITLE
[enh] Add engine for discourse.org forums

### PR DESCRIPTION
**Pull request by allendema**
## What does this PR do?
Add engine for forums based on [https://discourse.org](https://discourse.org).

## Why is this change important?
A lot of forums can be searched using this engine.

[https://trends.builtwith.com/websitelist/Discourse](https://trends.builtwith.com/websitelist/Discourse)

## How to test this PR locally?

Search for something (by default on [discuss.python.org](discuss.python.org)) ```!dpy lru_cache```
## Author's checklist

N/A
## Related issues

N/A

### CODE REVIEW

File: searx/engines/discourse.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        }
+    )
+
+    params['url'] = base_url + search_endpoint + args
+    params['headers'].update(
+        {
+            'Accept': 'application/json, text/javascript, */*; q=0.01',
+            'X-Requested-With': 'XMLHttpRequest',
+        }
+    )
```
Comment (Bnyro) [inline searx/engines/discourse.py]:
```suggestion
    params['headers'] = {
        'Accept': 'application/json, text/javascript, */*; q=0.01',
        'X-Requested-With': 'XMLHttpRequest',
    )
```
params['headers'] should be empty by default, shouldn't it?

File: searx/engines/discourse.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        content = '@' + post.get('username', '')
+
+        if int(comments) > 1:
+            content += f' | Comments: {comments}'
+
+        if result.get('has_accepted_answer', ''):
+            content += ' | Answered'
+        else:
+            if int(comments) > 1:
+                content += ' | No Accepted Answer'
```
Comment (Bnyro) [inline searx/engines/discourse.py]:
```suggestion
        status = "Closed" if result.get('closed', '') else "Open"
        comments = result.get('posts_count', 1)

        url = f"{base_url}/t/{result['id']}"
        publishedDate = parser.parse(result['created_at'])

        content = ['@' + post.get('username', '')]

        if int(comments) > 1:
            content += f'Comments: {comments}'

        if result.get('has_accepted_answer', ''):
            content += 'Answered'
        elif int(comments) > 1:
            content += 'No Accepted Answer'
            
        content = " | ".join(conent)
```
(not better performance-wise, but much easier to read)

File: searx/engines/discourse.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+        results.append(
+            {
+                'url': url,
+                'title': html.unescape(str(result['title']) + status),
```
Comment (Bnyro) [inline searx/engines/discourse.py]:
```suggestion
                'title': html.unescape(str(result['title']) + ' | ' + status),
```
(when following the suggestion above)

File: searx/engines/discourse.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                'content': html.unescape(content),
```
Comment (Bnyro) [inline searx/engines/discourse.py]: Why are we calling `html.unescape` here? All parameters that were added to `content` are self-created fields and not taken directly from the response?

File: searx/engines/discourse.py (hunk 2 above)
Comment (allendema) [reply searx/engines/discourse.py]: This joined every char, so I now used a list to append.
Thanks.

File: searx/engines/discourse.py (hunk 4 above)
Comment (allendema) [reply searx/engines/discourse.py]: Undone. :)

**Review by Bnyro — APPROVED** [review]

File: searx/settings.yml [hunk 5]
```diff
@@ -1581,6 +1581,13 @@ engines:
     api_site: 'superuser'
     categories: [it, q&a]
 
+  - name: discusspython
+    engine: discourse
+    shortcut: dpy
+    base_url: 'https://discuss.python.org'
+    categories: [it, q&a]
```
Comment (return42) [inline searx/settings.yml:1588]: I think we can better set these two values as defaults in the python module / can be modified in custom settings anyway .. but there should not be the need to set the defaults here in the settings.yml

@allendema can you patch it .. or if it is OK; I can patch it in my review ..

Comment (allendema) [reply searx/settings.yml:1588]: Maybe leave it so the admins can see an example of what is possible to do i.e use another forum like:
[https://caddy.community](https://caddy.community) or [https://discourse.pi-hole.net](https://discourse.pi-hole.net).

But you can patch it how you see fit.

File: searx/engines/discourse.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+def response(resp):
+
+    results = []
+    json_data = resp.json()
+
+    if ('topics' or 'posts') not in json_data.keys():
+        return []
+
+    for post, result in zip(json_data['posts'], json_data['topics']):
```
Comment (return42) [inline searx/engines/discourse.py]: python's `zip` function won't work here .. does not what we expect:

- `topics` and `posts` list do not have the same number of items
- the sorting of both lists is not the same / in the example below, `posts[4]` refers to topic id found in `topics[1]`

.. I will fix it in my review / give me some time, its in work ..

image

**Review by return42 — APPROVED** [review]
@allendema @Bnyro can you please review: I made some modifications mainly to

- fix issue reported [here](https://github.com/searxng/searxng/pull/3501#pullrequestreview-2082651373)
- add time-range support
- add online documentation
- separated commit for engine config (settings.yml) from commit with engine implementation.

Here is the [diff](https://github.com/searxng/searxng/compare/7f5146b0526c772901a1e83b847be528f4af207b..95fd8284c15cf321081f3070a01f883de49ed435) of my modification to what you have already been reviewed.

Sorry for the significant changes, but I think the engine benefits from the revision.

@allendema thank you for your contribution, it was a very interesting task for me / I didn't know Discourse yet.

File: searx/engines/discourse.py (hunk 6 above)
Comment (allendema) [reply searx/engines/discourse.py]: Thats right. Good catch.

**Comment by allendema** [comment]
@return42 Thanks for you fixes and enhancements. Tested it and everything works.

**Pull request merged**